In [1]:
import numpy as np
import statsmodels.formula.api as smf
import matplotlib as plt
import pandas as pd
%matplotlib inline

## Ejemplo 17.1 Participación en el mercado laboral de las mujeres casadas (mroz.dta)

In [2]:
df1 = pd.read_stata('mroz.dta')
df1.head()

inlf  hours  kidslt6  kidsge6  age  educ    wage  repwage  hushrs  husage  \
0     1   1610        1        0   32    12  3.3540     2.65    2708      34   
1     1   1656        0        2   30    12  1.3889     2.65    2310      30   
2     1   1980        1        3   35    12  4.5455     4.04    3072      40   
3     1    456        0        3   34    12  1.0965     3.25    1920      53   
4     1   1568        1        2   31    14  4.5918     3.60    2000      32   

   ...   faminc     mtr  motheduc  fatheduc  unem  city  exper   nwifeinc  \
0  ...  16310.0  0.7215        12         7   5.0     0     14  10.910060   
1  ...  21800.0  0.6615         7         7  11.0     1      5  19.499981   
2  ...  21040.0  0.6915        12         7   5.0     0     15  12.039910   
3  ...   7300.0  0.7815         7         7   5.0     0      6   6.799996   
4  ...  27300.0  0.6215        12        14   9.5     1      7  20.100058   

      lwage  expersq  
0  1.210154      196  
1  0.328512       25  
2  1.514138      225  
3  0.092123       36  
4  1.524272       49  

[5 rows x 22 columns]

In [7]:
reg1 = smf.ols('inlf ~ nwifeinc + educ + exper + expersq +age+kidslt6+ kidsge6', data = df1).fit()
print(reg1.summary())

                            OLS Regression Results                            
Dep. Variable:                   inlf   R-squared:                       0.264
Model:                            OLS   Adj. R-squared:                  0.257
Method:                 Least Squares   F-statistic:                     38.22
Date:                Tue, 08 Oct 2019   Prob (F-statistic):           6.90e-46
Time:                        21:17:53   Log-Likelihood:                -423.89
No. Observations:                 753   AIC:                             863.8
Df Residuals:                     745   BIC:                             900.8
Df Model:                           7                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept      0.5855      0.154      3.798      0.0

In [ ]:
# Como ejemplo de interpretación, si seaumenta la educación en un año, , ceteris paribuis, la probabilidad para la mujer 
# de participar en el mercado de trabajo , aumenta en 0,038
# El aumento de una unidad de otros ingresos (nwideinc) la probabilidad de la mujer para participar en el mercado de 
# trabajo disminuye en 0,0034, ceteris paribus,  que es un coeficiente bajo , pero es estadísticamente significativo
# Al aumentar la experiencia en un año, la probabildiad de la mujer para participar en el mercado de trabajo aumenta 
# en 0,039 – 2 * (0,0006)exper
# El punto en que la experiencia no afecta la probabilidad  participar en el mercado de trabajo, que es cuando cambia de 
# pendiente, es 0,039/0,0012 = 32,5 . Solo hay 13 mujeres con  exper > 32
# Al aumentar la edad un año, ceteris paribus, la probabilidad de la mujer de participar enel mercado de trabajo se reduce 
# en  0,016
# Tener un hijo  adicional de menos de 6 años, ceteris paribus, reduce la probabilidad de participar en 0,262
# Tener un hijo entre 6 y 18 años adicional, ceteris paribus, aumeta la probabilidad de la mujer de participar en el mercado
# de trabajo en 0,013


In [8]:
reg2 = smf.logit('inlf ~ nwifeinc + educ + exper + expersq +age+kidslt6+ kidsge6', data = df1).fit()
print(reg2.summary())

Optimization terminated successfully.
         Current function value: 0.533553
         Iterations 6
                           Logit Regression Results                           
Dep. Variable:                   inlf   No. Observations:                  753
Model:                          Logit   Df Residuals:                      745
Method:                           MLE   Df Model:                            7
Date:                Tue, 08 Oct 2019   Pseudo R-squ.:                  0.2197
Time:                        21:17:53   Log-Likelihood:                -401.77
converged:                       True   LL-Null:                       -514.87
                                        LLR p-value:                 3.159e-45
                 coef    std err          z      P>|z|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept      0.4255      0.860      0.494      0.621      -1.261       2.112
nwifeinc      -0.0213      0.

In [29]:
from scipy.stats import logistic
logistic.pdf(reg2.fittedvalues)

array([0.20973456, 0.18800195, 0.20918232, 0.17372607, 0.24341655,
       0.15155625, 0.06921232, 0.15531693, 0.12797552, 0.07479226,
       0.0976142 , 0.19064808, 0.17746039, 0.16185309, 0.23546894,
       0.17971209, 0.10745401, 0.21939012, 0.11453204, 0.13822804,
       0.20941053, 0.09352016, 0.16001609, 0.16899973, 0.23847946,
       0.05082137, 0.20589884, 0.13264094, 0.19865055, 0.2431598 ,
       0.13007686, 0.24459843, 0.07323989, 0.23204923, 0.04180409,
       0.18908936, 0.15884414, 0.1309055 , 0.11782347, 0.04301659,
       0.19545411, 0.24871709, 0.23177698, 0.2447791 , 0.15493556,
       0.20780808, 0.18256281, 0.10753179, 0.22010665, 0.14489354,
       0.23361218, 0.19454972, 0.08048576, 0.24999943, 0.17382002,
       0.06626962, 0.13233766, 0.07059131, 0.13898737, 0.22715783,
       0.0797646 , 0.20048991, 0.24156616, 0.19029938, 0.14173264,
       0.21575675, 0.09351909, 0.13618627, 0.21429134, 0.15961958,
       0.11011316, 0.24965376, 0.24168628, 0.1647638 , 0.07748

In [30]:
np.mean(logistic.pdf(reg2.fittedvalues))

0.17857963439562646

In [10]:
margeff = reg2.get_margeff()
print(margeff.summary())

        Logit Marginal Effects       
Dep. Variable:                   inlf
Method:                          dydx
At:                           overall
                dy/dx    std err          z      P>|z|      [0.025      0.975]
------------------------------------------------------------------------------
nwifeinc      -0.0038      0.001     -2.571      0.010      -0.007      -0.001
educ           0.0395      0.007      5.414      0.000       0.025       0.054
exper          0.0368      0.005      7.139      0.000       0.027       0.047
expersq       -0.0006      0.000     -3.176      0.001      -0.001      -0.000
age           -0.0157      0.002     -6.603      0.000      -0.020      -0.011
kidslt6       -0.2578      0.032     -8.070      0.000      -0.320      -0.195
kidsge6        0.0107      0.013      0.805      0.421      -0.015       0.037


In [11]:
reg3 = smf.probit('inlf ~ nwifeinc + educ + exper + expersq +age+kidslt6+ kidsge6', data = df1).fit()
print(reg3.summary())

Optimization terminated successfully.
         Current function value: 0.532938
         Iterations 5
                          Probit Regression Results                           
Dep. Variable:                   inlf   No. Observations:                  753
Model:                         Probit   Df Residuals:                      745
Method:                           MLE   Df Model:                            7
Date:                Tue, 08 Oct 2019   Pseudo R-squ.:                  0.2206
Time:                        21:19:22   Log-Likelihood:                -401.30
converged:                       True   LL-Null:                       -514.87
                                        LLR p-value:                 2.009e-45
                 coef    std err          z      P>|z|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept      0.2701      0.509      0.531      0.595      -0.727       1.267
nwifeinc      -0.0120      0.

In [ ]:
# Para cuantificar e interpretar los efectos marginales, sobre la probabilidad de participación en el mercado de trabajo de la
# mujer, debemos tener presente que en el modelo Prbit el efcto marginal de la variable  Xj es 
# dP(Y = 1|X)/dXj = g(XB) Bj
# Es decir, es el coeficiente estimado Bj multiplicado por el valor de la normal en la regresión estimada para  los 
# valores específicos de las variables  explicativas
# Es decir que tenemos 753 efectos marginales, uno para cada individuo de la muestra

In [ ]:
# Para crear la serie de lso efectos marginales de cada individuo lo que tenemos que hace es calcular la densidad de la normal ,
# para la estimación 

In [18]:
reg3.params[0]+ reg3.params[1]*df1.nwifeinc +df1.educ*reg3.params[2]+ df1.exper*reg3.params[3] +df1.expersq*reg3.params[4]+ df1.age*reg3.params[5]+df1.kidslt6* reg3.params[6]+ df1.kidsge6*reg3.params[7]

0      0.507138
1      0.662462
2      0.511633
3      0.742347
4      0.197280
5      0.883787
6      1.499459
7      0.861323
8      1.041328
9      1.454380
10     1.266245
11     0.646352
12    -0.721321
13     0.821731
14    -0.281541
15     0.714687
16     1.182864
17     0.441411
18     1.131758
19     0.968541
20     0.513914
21     1.286568
22     0.833590
23     0.773237
24    -0.266122
25     1.724050
26     0.540537
27     1.010534
28     0.578103
29     0.203315
         ...   
723    1.205732
724   -0.900731
725   -0.644968
726   -0.589577
727   -0.277831
728   -0.657508
729   -0.275200
730   -0.008761
731   -0.669516
732    0.738655
733    1.194505
734   -0.196259
735   -0.126004
736    0.656693
737    0.584075
738   -0.105116
739   -0.528759
740    0.366850
741   -0.365979
742   -0.279881
743   -0.077434
744    0.511708
745   -1.265653
746   -0.751095
747    0.409358
748    0.160233
749   -0.189380
750   -0.088194
751   -0.205007
752    0.362382
Length: 753, dtype: floa

In [19]:
reg3.fittedvalues

0      0.507138
1      0.662462
2      0.511633
3      0.742347
4      0.197280
5      0.883787
6      1.499459
7      0.861323
8      1.041328
9      1.454380
10     1.266245
11     0.646352
12    -0.721321
13     0.821731
14    -0.281541
15     0.714687
16     1.182864
17     0.441411
18     1.131758
19     0.968541
20     0.513914
21     1.286568
22     0.833590
23     0.773237
24    -0.266122
25     1.724050
26     0.540537
27     1.010534
28     0.578103
29     0.203315
         ...   
723    1.205732
724   -0.900731
725   -0.644968
726   -0.589577
727   -0.277831
728   -0.657508
729   -0.275200
730   -0.008761
731   -0.669516
732    0.738655
733    1.194505
734   -0.196259
735   -0.126004
736    0.656693
737    0.584075
738   -0.105116
739   -0.528759
740    0.366850
741   -0.365980
742   -0.279881
743   -0.077434
744    0.511708
745   -1.265653
746   -0.751095
747    0.409358
748    0.160233
749   -0.189380
750   -0.088194
751   -0.205007
752    0.362382
Length: 753, dtype: floa

In [21]:
from scipy.stats import norm
norm.pdf(reg3.fittedvalues)

array([0.35080203, 0.32034196, 0.34999988, 0.30286201, 0.39125403,
       0.2699609 , 0.12962281, 0.27530468, 0.23197608, 0.13854661,
       0.17895398, 0.32373702, 0.30755825, 0.28463173, 0.38344037,
       0.3090269 , 0.1981914 , 0.36190979, 0.21026714, 0.24958043,
       0.34959074, 0.1743716 , 0.28185162, 0.29585481, 0.38506278,
       0.09025528, 0.34471789, 0.2394218 , 0.33755053, 0.39078137,
       0.23698705, 0.39299403, 0.13579677, 0.3774447 , 0.07237266,
       0.3241285 , 0.28101021, 0.23682033, 0.21661496, 0.07504805,
       0.33202262, 0.39769328, 0.3776236 , 0.39250457, 0.27428417,
       0.34866801, 0.31442274, 0.19956963, 0.36281415, 0.26053149,
       0.37832979, 0.33002164, 0.14898143, 0.39893438, 0.30182584,
       0.12317706, 0.24004431, 0.1308542 , 0.24964898, 0.37186876,
       0.14632107, 0.33794089, 0.38962729, 0.3249627 , 0.25578178,
       0.35659959, 0.17494131, 0.24534318, 0.354821  , 0.28071837,
       0.20306492, 0.39832749, 0.38881952, 0.28861066, 0.14443

In [22]:
margeffP = reg3.get_margeff()
print(margeffP.summary())

       Probit Marginal Effects       
Dep. Variable:                   inlf
Method:                          dydx
At:                           overall
                dy/dx    std err          z      P>|z|      [0.025      0.975]
------------------------------------------------------------------------------
nwifeinc      -0.0036      0.001     -2.509      0.012      -0.006      -0.001
educ           0.0394      0.007      5.452      0.000       0.025       0.054
exper          0.0371      0.005      7.200      0.000       0.027       0.047
expersq       -0.0006      0.000     -3.205      0.001      -0.001      -0.000
age           -0.0159      0.002     -6.739      0.000      -0.021      -0.011
kidslt6       -0.2612      0.032     -8.197      0.000      -0.324      -0.199
kidsge6        0.0108      0.013      0.829      0.407      -0.015       0.036


In [28]:
margeffP1 = reg3.get_margeff(at='mean')
print(margeffP1.summary())

       Probit Marginal Effects       
Dep. Variable:                   inlf
Method:                          dydx
At:                              mean
                dy/dx    std err          z      P>|z|      [0.025      0.975]
------------------------------------------------------------------------------
nwifeinc      -0.0047      0.002     -2.484      0.013      -0.008      -0.001
educ           0.0511      0.010      5.186      0.000       0.032       0.070
exper          0.0482      0.007      6.575      0.000       0.034       0.063
expersq       -0.0007      0.000     -3.141      0.002      -0.001      -0.000
age           -0.0206      0.003     -6.241      0.000      -0.027      -0.014
kidslt6       -0.3392      0.046     -7.316      0.000      -0.430      -0.248
kidsge6        0.0141      0.017      0.828      0.408      -0.019       0.047


## Ejemplo 17.2 La oferta laboral anual de las mujeres casadas* (mroz.dta)

Tobit

El archivo MROZ incluye datos sobre las horas trabajadas de 753 mujeres casadas, 428 de las cuales trabajaron por un salario fuera del hogar durante el año; 325 de las mujeres trabajaron cero horas. Para las mujeres que trabajaron horas positivas, el rango es muy amplio, que va de 12 a 4,950. 

Por tanto, las horas anuales trabajadas es un buen candidato para un modelo Tobit. También se estima un modelo lineal
(usando todas las 753 observaciones) mediante MCO. Los resultados se dan en la tabla 17.2.

Esta tabla tiene varias características notorias. Primero, las estimaciones Tobit de los coeficientes tienen el mismo signo que las estimaciones de MCO correspondientes y su significancia estadística es similar.

(Las posibles excepciones son los coeficientes de nwifeinc y kidsge6, pero los estadísticos t tienen magnitudes similares.) Segundo, aunque es tentador comparar las magnitudes de las estimaciones de MCO y de Tobit, esto no es muy informativo. Es necesario tener cuidado en no pensar que, debido a que el coeficiente Tobit en kidslt6 es aproximadamente el doble que el coeficiente de MCO, el modelo Tobit implica una respuesta mucho mayor de horas trabajadas respecto a los niños pequeños.

Se pueden multiplicar las estimaciones Tobit mediante los factores adecuados de ajuste para hacerlas aproximadamente comparables a las estimaciones de MCO. El factor escalar EPP $n^{-1} \sum_{i-1}^{n} \Phi(x_{i}\hat \beta /\hat \sigma)$ resulta ser de cerca de 0.589, el cual se puede usar para obtener los efectos parciales promedio para la estimación Tobit. 

Si, por ejemplo, se multiplica el coeficiente de educ por 0.589 se obtiene 0.589(80.65) $\sim$ 47.50 (es decir, 47.5 horas más), lo cual es un poco mayor que el efecto parcial de MCO, de aproximadamente 28.8 horas. Así que, incluso para estimar un efecto promedio, las estimaciones Tobit son notablemente mayores en magnitud que la estimación de MCO correspondiente. 

Si, en lugar de ello, se desea el efecto estimado de otro año de educación a partir de los valores promedio de todas las variables explicativas, entonces se
calcula el factor escalar EPA $\Phi(\bar x_{i}\hat \beta /\hat \sigma)$. Éste resulta ser de aproximadamente 0.645 [cuando se usa el promedio cuadrado de la experiencia, $(\overline exper)^{2}$, en lugar del promedio de $exper^{2}$]. Este efecto parcial, que es de aproximadamente 52 horas, es casi el doble de la estimación de MCO. Con excepción de kidsge6, los coeficientes de pendiente escalados de Tobit son todos mayores en magnitud que el coeficiente correspondiente de MCO.

Se ha reportado una R-cuadrada tanto para el modelo Tobit como para el de regresión lineal. La R-cuadrada para MCO es la acostumbrada. Para Tobit, la R-cuadrada es el coeficiente de correlación entre $y_{i}$ y $\hat y_{i}$, donde $\hat y{i} = \Phi(x_{i} \hat \beta /\hat \sigma) x_{i} \hat \beta + \hat \sigma \phi(x_{i} \hat \beta /\hat \sigma)$ es la estimación de $E(y|x = x_{i})$. Esto está motivado por el hecho de que la R-cuadrada acostumbrada para MCO es igual a la correlación cuadrada entre $y_{i}$ y los valores ajustados [vea la ecuación (3.29)]. En los modelos no lineales como el modelo Tobit, el coeficiente de correlación cuadrada no es idéntico a una R-cuadrada basado en una suma de residuales cuadrados como en
(3.28). Esto es debido a que los valores ajustados, como se definieron antes, y los residuales, $y_{i} - \hat y_{i}$, están
correlacionados en la muestra. Una R-cuadrada definida como el coeficiente de correlación cuadrado entre
$y_{i}$ y $\hat y_{i}$ tiene la ventaja de siempre estar entre cero y uno; una R-cuadrada basada en la suma de los residuales
cuadrados no necesita tener esta característica.

Se puede ver que, con base en las medidas de R-cuadrada, la función de media condicional Tobit encaja un poco mejor con los datos de las horas, pero no es una mejora sustancial. No obstante, se debe recordar que las estimaciones Tobit no se eligen para maximizar una R-cuadrada (maximizan la función de log-verosimilitud), mientras que las estimaciones de MCO son los valores que producen la R-cuadrada más alta dada la forma funcional lineal.

Por construcción, todos los valores ajustados Tobit para hours son positivos. En contrate, 39 de los
valores ajustados por MCO son negativos. Aunque las predicciones negativas son de algún interés, 39 de
753 sólo está por encima de 5% de las observaciones. No es completamente claro cómo se traducen los
valores ajustados negativos de MCO a las diferencias en los efectos parciales estimados. 

La figura 17.3 presenta la gráfica de las estimaciones de $E(hours|x) en función de la educación; para el modelo Tobit,
las demás variables explicativas se establecen en sus valores promedio. Para el modelo lineal, la ecuación
graficada es 

$\widehat{hours} = 387.19 + 28.76 educ$. Para el modelo Tobit, la ecuación graficada es $\widehat{hours} = \Phi[(-694.12 + 80.65 educ)/1,122.02] (-694.12 + 80.65 educ) + 1,122.02 \phi[(-694.12 + 80.65 educ)/1,122.02]$. 

Como se puede ver en la figura, el modelo lineal da estimaciones notablemente superiores de las horas trabajadas esperadas incluso para niveles muy altos de educación. 

Por ejemplo, a los ocho años de educación, el valor predicho por MCO de las horas es de cerca de 617.5, mientras que la estimación Tobit es de aproximadamente 423.9. A los 12 años de educación, las horas predichas son de cerca de 732.7
y 598.3, respectivamente. Las dos líneas de predicción cruzan después de los 17 años de educación, pero ninguna mujer en la muestra tiene más de 17 años de educación. La pendiente creciente de la línea Tobit indica con claridad el creciente efecto marginal de la educación sobre las horas trabajadas esperadas.

In [3]:
regEj172 = smf.ols('hours ~ nwifeinc + educ + exper + expersq +age+kidslt6+ kidsge6', data = df1).fit()
print(regEj172.summary())

                            OLS Regression Results                            
Dep. Variable:                  hours   R-squared:                       0.266
Model:                            OLS   Adj. R-squared:                  0.259
Method:                 Least Squares   F-statistic:                     38.50
Date:                Tue, 10 Oct 2023   Prob (F-statistic):           3.42e-46
Time:                        01:36:09   Log-Likelihood:                -6049.5
No. Observations:                 753   AIC:                         1.212e+04
Df Residuals:                     745   BIC:                         1.215e+04
Df Model:                           7                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept   1330.4824    270.785      4.913      0.0

In [4]:
from py4etrics.truncreg import Truncreg
from py4etrics.tobit import Tobit
from py4etrics.heckit import Heckit

In [6]:
left = 0

In [7]:
censor = df1['hours'].apply(lambda x: -1 if x==left else 0)

In [8]:
regEj172T = Tobit.from_formula('hours ~1+ nwifeinc + educ + exper + expersq +age+kidslt6+ kidsge6',cens=censor,left=0, data = df1).fit()
print(regEj172T.summary())

Optimization terminated successfully.
         Current function value: 5.071839
         Iterations: 2418
         Function evaluations: 3520
                              Tobit Regression Results                             
Dep. Variable:                       hours   Pseudo R-squ:                    0.034
Method:                 Maximum Likelihood   Log-Likelihood:                -3819.1
No. Observations:                      753   LL-Null:                       -3954.9
No. Uncensored Obs:                    428   LL-Ratio:                        271.6
No. Left-censored Obs:                 325   LLR p-value:                     0.000
No. Right-censored Obs:                  0   AIC:                            7654.2
Df Residuals:                          745   BIC:                            7691.2
Df Model:                                7   Covariance Type:             nonrobust
                 coef    std err          z      P>|z|      [0.025      0.975]
-----------------------

## Ejemplo 17.3 Modelo de regresión de Poisson para el número de detenciones (crime1.dta)

In [14]:
dfej17_3 = pd.read_stata('crime1.dta')
dfej17_3.head()

narr86  nfarr86  nparr86  pcnv     avgsen    tottime  ptime86  qemp86  \
0       0        0        0  0.38  17.600000  35.200001       12     0.0   
1       2        2        0  0.44   0.000000   0.000000        0     1.0   
2       1        1        0  0.33  22.799999  22.799999        0     0.0   
3       2        2        1  0.25   0.000000   0.000000        5     2.0   
4       1        1        0  0.00   0.000000   0.000000        0     2.0   

   inc86  durat  black  hispan  born60  pcnvsq  pt86sq    inc86sq  
0    0.0    0.0      0       0       1  0.1444     144   0.000000  
1    0.8    0.0      0       1       0  0.1936       0   0.640000  
2    0.0   11.0      1       0       1  0.1089       0   0.000000  
3    8.8    0.0      0       1       1  0.0625      25  77.440002  
4    8.1    1.0      0       0       0  0.0000       0  65.610008

In [15]:
#import wooldridge as woo
import pandas as pd
import statsmodels.api as sm
import statsmodels.formula.api as smf

#crime1 = woo.dataWoo('crime1')

# estimate linear model:
reg_lin = smf.ols(formula='narr86 ~ pcnv + avgsen + tottime + ptime86 +'
                          'qemp86 + inc86 + black + hispan + born60',
                  data=dfej17_3)
results_lin = reg_lin.fit()

# print regression table:
table_lin = pd.DataFrame({'b': round(results_lin.params, 4),
                          'se': round(results_lin.bse, 4),
                          't': round(results_lin.tvalues, 4),
                          'pval': round(results_lin.pvalues, 4)})
print(f'table_lin: \n{table_lin}\n')

# estimate Poisson model:
reg_poisson = smf.poisson(formula='narr86 ~ pcnv + avgsen + tottime +'
                                  'ptime86 + qemp86 + inc86 + black +'
                                  'hispan + born60',
                          data=dfej17_3)
results_poisson = reg_poisson.fit(disp=0)

# print regression table:
table_poisson = pd.DataFrame({'b': round(results_poisson.params, 4),
                              'se': round(results_poisson.bse, 4),
                              't': round(results_poisson.tvalues, 4),
                              'pval': round(results_poisson.pvalues, 4)})
print(f'table_poisson: \n{table_poisson}\n')

# estimate Quasi-Poisson model:
reg_qpoisson = smf.glm(formula='narr86 ~ pcnv + avgsen + tottime + ptime86 +'
                               'qemp86 + inc86 + black + hispan + born60',
                       family=sm.families.Poisson(),
                       data=dfej17_3)
# the argument scale controls for the dispersion in exponential dispersion models,
# see the module documentation for more details:
results_qpoisson = reg_qpoisson.fit(scale='X2', disp=0)

# print regression table:
table_qpoisson = pd.DataFrame({'b': round(results_qpoisson.params, 4),
                               'se': round(results_qpoisson.bse, 4),
                               't': round(results_qpoisson.tvalues, 4),
                               'pval': round(results_qpoisson.pvalues, 4)})
print(f'table_qpoisson: \n{table_qpoisson}\n')


table_lin: 
                b      se        t    pval
Intercept  0.5766  0.0379  15.2150  0.0000
pcnv      -0.1319  0.0404  -3.2642  0.0011
avgsen    -0.0113  0.0122  -0.9257  0.3547
tottime    0.0121  0.0094   1.2790  0.2010
ptime86   -0.0409  0.0088  -4.6378  0.0000
qemp86    -0.0513  0.0145  -3.5420  0.0004
inc86     -0.0015  0.0003  -4.2613  0.0000
black      0.3270  0.0454   7.1987  0.0000
hispan     0.1938  0.0397   4.8799  0.0000
born60    -0.0225  0.0333  -0.6747  0.4999

table_poisson: 
                b      se       t    pval
Intercept -0.5996  0.0673 -8.9158  0.0000
pcnv      -0.4016  0.0850 -4.7260  0.0000
avgsen    -0.0238  0.0199 -1.1918  0.2333
tottime    0.0245  0.0148  1.6603  0.0969
ptime86   -0.0986  0.0207 -4.7625  0.0000
qemp86    -0.0380  0.0290 -1.3099  0.1902
inc86     -0.0081  0.0010 -7.7624  0.0000
black      0.6608  0.0738  8.9503  0.0000
hispan     0.4998  0.0739  6.7609  0.0000
born60    -0.0510  0.0641 -0.7967  0.4256

table_qpoisson: 
                b 

## Ejemplo 17.4 La duración hasta la reincidencla (recid.dta)

Censurada

El archivo RECID.RAW contiene datos sobre el tiempo en meses hasta que un recluso en una prisión de
Carolina del Norte es arrestado después de haber sido liberado de prisión: llamemos a esto durat. Algunos
reclusos participaron en un programa laboral mientras estaban en prisión. También se controla una variedad
de variables demográficas, así como medidas de historia criminal y de prisión.
De 1,445 reclusos, 893 no habían sido arrestados durante el periodo de seguimiento; por tanto, estas
observaciones están censuradas. Las duraciones censuradas difirieron entre los reclusos, con un rango de
70 a 81 meses.
La tabla 17.4 da los resultados de la regresión censurada normal para log(durat). Cada uno de los
coeficientes, multiplicados por 100, dan el cambio porcentual estimado en la duración esperada dado un
incremento ceteris paribus de una unidad en la variable explicativa correspondiente.
Varios coeficientes en la tabla 17.4 son interesantes. Las variables priors (número de condenas anteriores)
y tserved (meses totales pasados en prisión) tienen efectos negativos sobre el periodo que transcurre
hasta que ocurra un nuevo arresto. Esto sugiere que estas variables miden la tendencia de la actividad criminal
más que representar un efecto disuasivo. Por ejemplo, un recluso con una condena previa adicional
tiene un periodo hasta el siguiente arresto que es casi 14% menor. Un año adicional de tiempo en prisión
reduce la duración por casi 100  12(.019)  22.8%. Un hallazgo un tanto sorprendente es que un hombre
que purga una condena por un delito grave (felon) tiene un periodo esperado estimado que es casi 56%
[exp(.444)  1  .56] mayor que un hombre que purga una condena por un delito no grave.
Aquellos que tienen una historia de abuso de drogas y alcohol tienen duraciones esperadas sustancialmente
menores hasta el siguiente arresto. (Las variables alcohol y drugs son variables binarias.) Los
hombres de mayor edad (age) y los hombres que estaban casados (married) al momento de su condena en
prisión, tienen periodos esperados más largos hasta su siguiente arresto. Los hombres negros (black) tienen
periodos sustancialmente menores del orden de 42% [exp(.543)  1  .42].
La variable de política clave, workprg, no tiene el efecto deseado. La estimación puntual es que, todo
lo demás constante, los hombres que participaron en el programa laboral tienen periodos de reincidencia
que son aproximadamente 6.3% menores que los hombres que no participaron. El coeficiente tiene un
estadístico t pequeño, así que probablemente concluiríamos que el programa de trabajo no tiene efecto. Esto
se puede deber a un problema de autoselección, o podría ser producto de la forma en que los hombres se
reclutaron en el programa. Por supuesto, simplemente puede deberse a que el programa fue ineficaz.

In [16]:
dfej17_4 = pd.read_stata('recid.dta')
dfej17_4.head()

black  alcohol  drugs  super  married  felon  workprg  property  person  \
0      0        1      0      1        1      0        1         0       0   
1      1        0      0      1        0      1        1         1       0   
2      0        0      0      0        0      0        1         1       0   
3      0        0      1      1        0      1        1         1       0   
4      0        0      1      1        0      0        0         0       0   

   priors  educ  rules  age  tserved  follow  durat  cens    ldurat  
0       0     7      2  441       30      72     72     1  4.276666  
1       0    12      0  307       19      75     75     1  4.317488  
2       0     9      5  262       27      81      9     0  2.197225  
3       2     9      3  253       38      76     25     0  3.218876  
4       0     9      0  244        4      81     81     1  4.394449

In [18]:
#import wooldridge as woo
import numpy as np
import patsy as pt
import scipy.stats as stats
import statsmodels.formula.api as smf
import statsmodels.base.model as smclass

#recid = woo.dataWoo('recid')

# define dummy for censored observations:
censored = dfej17_4['cens'] != 0
y, X = pt.dmatrices('ldurat ~ workprg + priors + tserved + felon +'
                    'alcohol + drugs + black + married + educ + age',
                    data=dfej17_4, return_type='dataframe')

# generate starting solution:
reg_ols = smf.ols(formula='ldurat ~ workprg + priors + tserved + felon +'
                          'alcohol + drugs + black + married + educ + age',
                  data=dfej17_4)
results_ols = reg_ols.fit()
sigma_start = np.log(sum(results_ols.resid ** 2) / len(results_ols.resid))
params_start = np.concatenate((np.array(results_ols.params), sigma_start),
                              axis=None)


# extend statsmodels class by defining nloglikeobs:
class CensReg(smclass.GenericLikelihoodModel):
    def __init__(self, endog, cens, exog):
        self.cens = cens
        super(smclass.GenericLikelihoodModel, self).__init__(endog, exog,
                                                             missing='none')

    def nloglikeobs(self, params):
        X = self.exog
        y = self.endog
        cens = self.cens
        p = X.shape[1]
        beta = params[0:p]
        sigma = np.exp(params[p])
        y_hat = np.dot(X, beta)
        ll = np.empty(len(y))
        # uncensored:
        ll[~cens] = np.log(stats.norm.pdf((y - y_hat)[~cens] /
                                          sigma)) - np.log(sigma)
        # censored:
        ll[cens] = np.log(stats.norm.cdf(-(y - y_hat)[cens] / sigma))
        return -ll


# results of MLE:
reg_censReg = CensReg(endog=y, exog=X, cens=censored)
results_censReg = reg_censReg.fit(start_params=params_start,
                                  maxiter=10000, method='BFGS', disp=0)
print(f'results_censReg.summary(): \n{results_censReg.summary()}\n')


c:\python\lib\site-packages\ipykernel_launcher.py:44: RuntimeWarning: divide by zero encountered in log
c:\python\lib\site-packages\statsmodels\tools\numdiff.py:158: RuntimeWarning: invalid value encountered in double_scalars
  f(*((x-ei,)+args), **kwargs))/(2 * epsilon[k])
c:\python\lib\site-packages\ipykernel_launcher.py:44: RuntimeWarning: divide by zero encountered in log


results_censReg.summary(): 
                               CensReg Results                                
Dep. Variable:                 ldurat   Log-Likelihood:                -1597.1
Model:                        CensReg   AIC:                             3216.
Method:            Maximum Likelihood   BIC:                             3274.
Date:                Wed, 04 Oct 2023                                         
Time:                        15:45:28                                         
No. Observations:                1445                                         
Df Residuals:                    1434                                         
Df Model:                          10                                         
                 coef    std err          z      P>|z|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept      4.0994      0.348     11.796      0.000       3.418       4.781
workprg       -0.0626   

c:\python\lib\site-packages\statsmodels\base\model.py:2653: UserWarning: df_model + k_constant differs from nparams
  warnings.warn("df_model + k_constant differs from nparams")
c:\python\lib\site-packages\statsmodels\base\model.py:2655: UserWarning: df_resid differs from nobs - nparams
  warnings.warn("df_resid differs from nobs - nparams")


## Ejemplo 17.5 Ecuación de oferta salarial para mujeres casadas* (mroz.dta)

In [3]:
import heckman as heckman


In [9]:
df1.shape

(753, 22)

In [10]:
df1['unos']=1

In [11]:
res = heckman.Heckman(df1.lwage, df1[['unos','educ', 'exper','expersq']],df1[['nwifeinc', 'age', 'kidslt6' , 'kidsge6']]).fit(method='twostep')
print(res.summary())

       Heckman Regression Results      
Dep. Variable:                    lwage
Model:                          Heckman
Method:                Heckman Two-Step
Date:                  Wed, 04 Oct 2023
Time:                          15:21:57
No. Total Obs.:                     753
No. Censored Obs.:                  325
No. Uncensored Obs.:                428
                 coef    std err          z      P>|z|      [0.025      0.975]
------------------------------------------------------------------------------
unos          -0.5337      0.236     -2.257      0.024      -0.997      -0.070
educ           0.1072      0.015      7.374      0.000       0.079       0.136
exper          0.0416      0.013      3.171      0.002       0.016       0.067
expersq       -0.0008      0.000     -2.071      0.038      -0.002   -4.34e-05
                 coef    std err          z      P>|z|      [0.025      0.975]
------------------------------------------------------------------------------
nwifeinc

In [ ]:
ver solomon y solucion con ratio mills como phi /Phi

In [13]:
import statsmodels.formula.api as smf
import scipy.stats as stats

#mroz = woo.dataWoo('mroz')

# step 1 (use all n observations to estimate a probit model of s_i on z_i):
reg_probit = smf.probit(formula='inlf ~ educ + exper + I(exper**2) +'
                                'nwifeinc + age + kidslt6 + kidsge6',
                        data=df1)
results_probit = reg_probit.fit(disp=0)
pred_inlf = results_probit.fittedvalues
df1['inv_mills'] = stats.norm.pdf(pred_inlf) / stats.norm.cdf(pred_inlf)

# step 2 (regress y_i on x_i and inv_mills in sample selection):
reg_heckit = smf.ols(formula='lwage ~ educ + exper + I(exper**2) + inv_mills',
                     subset=(df1['inlf'] == 1), data=df1)
results_heckit = reg_heckit.fit()

# print results:
print(f'results_heckit.summary(): \n{results_heckit.summary()}\n')


results_heckit.summary(): 
                            OLS Regression Results                            
Dep. Variable:                  lwage   R-squared:                       0.149
Model:                            OLS   Adj. R-squared:                  0.141
Method:                 Least Squares   F-statistic:                     18.58
Date:                Wed, 04 Oct 2023   Prob (F-statistic):           4.44e-14
Time:                        15:33:10   Log-Likelihood:                -433.47
No. Observations:                 428   AIC:                             876.9
Df Residuals:                     423   BIC:                             897.2
Df Model:                           4                                         
Covariance Type:            nonrobust                                         
                    coef    std err          t      P>|t|      [0.025      0.975]
---------------------------------------------------------------------------------
Intercept        -0

In [ ]:
con py4etrics


In [13]:
endog = df1.loc[:,'lwage']


In [9]:

exog = df1.loc[:,['educ', 'exper', 'expersq']]
exog['Intercept'] = 1.0

In [10]:
exog_select = df1.loc[:,['educ', 'exper', 'expersq','nwifeinc', 'age', 'kidslt6', 'kidsge6', ]]
exog_select['Intercept'] = 1.0

In [14]:
res_heckit = Heckit(endog, exog, exog_select).fit(cov_type_2='HC1')

print(res_heckit.summary())

                           Heckit Regression Results                            
Dep. Variable:                    lwage   R-squared:                       0.156
Model:                           Heckit   Adj. R-squared:                  0.150
Method:                Heckman Two-Step   F-statistics:                   26.148
Date:                  Tue, 10 Oct 2023   Prob (F-statistic):              0.000
Time:                          10:58:51   Cov in 1st Stage:            nonrobust
No. Total Obs.:                     753   Cov in 2nd Stage:                  HC1
No. Censored Obs.:                  325                                         
No. Uncensored Obs.:                428                                         
                 coef    std err          z      P>|z|      [0.025      0.975]
------------------------------------------------------------------------------
educ           0.1091      0.016      7.026      0.000       0.079       0.139
exper          0.0439      0.016  